File: wasm/lib/detailed/operation_factories/comment.ts
import type { TAccountName } from "../custom_jsons/builder";
import { comment } from "../../proto/comment.js";
import { beneficiary_route_type, comment_options, type comment_payout_beneficiaries } from "../../proto/comment_options.js";
import { NaiAsset } from "../index.js";
import { WaxError } from "../../errors.js";
import { IBuiltHiveAppsOperation, OperationBuilder } from "../operation_builder.js";

export enum ECommentFormat {
  HTML = "html",
  MARKDOWN = "markdown",
  MIXED = "markdown+html"
}

class CommentBuilder extends OperationBuilder {
  protected readonly comment: comment;
  private commentOptions?: comment_options;

  protected jsonMetadata: Record<string, any> = {
    format: ECommentFormat.MIXED
  };

  private extendDefaultJsonMetadata(optionalJsonMeta: { app?: string } = {}): void {
    const apps = [ `${process.env.npm_package_name}/${process.env.npm_package_version}` ];
    if(typeof optionalJsonMeta.app === "string")
      apps.unshift(optionalJsonMeta.app);

    Object.assign(this.jsonMetadata, optionalJsonMeta, { app: apps.join(", ") });
  }

  protected constructor(parentAuthor: TAccountName, parentPermlink: string, author: TAccountName, body: string, jsonMetadata?: object, permlink?: string, title: string = "") {
    super();

    if(permlink === undefined)
      permlink = `re-${parentAuthor}-${Date.now()}`;

    this.comment = comment.fromPartial({
      parent_author: parentAuthor,
      parent_permlink: parentPermlink,
      author,
      body,
      permlink,
      title
    });

    this.extendDefaultJsonMetadata(jsonMetadata);
  }

  private ensureCommentOptionsCreated(): void {
    this.requireApi();

    if(typeof this.commentOptions === "undefined")
      this.commentOptions = comment_options.fromPartial({
        author: this.comment.author,
        permlink: this.comment.permlink,
        allow_curation_rewards: true,
        allow_votes: true,
        percent_hbd: 10000,
        max_accepted_payout: this.api!.hbd(1000000000)
      });
  }

  /**
   * Assigns given object or sets given value on key in comment meta values
   *
   * @param {string | object} keyOrObject key to set the value on or the entire object of key-value pairs to assign to the json metadata object
   * @param {?any} value value to be set (optional when passing the entire object)
   *
   * @throws {WaxError} if key already exists on the jsonmetadata object
   * @returns {CommentBuilder} itself
   */
  public pushMetadataProperty(keyOrObject: string | object, value?: any): CommentBuilder {
    const assign = (key: string, value: any) => {
      if(key in this.jsonMetadata)
        throw new WaxError("Key already exists on the json metadata object");

      this.jsonMetadata[key] = value;
    }

    if(typeof keyOrObject === "string")
      assign(keyOrObject, value);
    else
      for(const key in keyOrObject)
        assign(key, keyOrObject[key]);

    return this;
  }

  /**
   * Pushes tags to the json metadata object
   *
   * @param {string[]} tags tags to be pushed to the json metadata object
   *
   * @returns {CommentBuilder} itself
   */
  public pushTags(...tags: string[]): CommentBuilder {
    if(tags.length === 0)
      return this;

    if(typeof this.jsonMetadata.tags === "undefined")
      this.jsonMetadata.tags = [];

    this.jsonMetadata.tags = [... new Set([ ...this.jsonMetadata.tags, ...tags ])];

    return this;
  }

  /**
   * Sets alternative author to the json metadata object
   *
   * @param {string} author alternative author to be set on the json metadata object
   *
   * @returns {CommentBuilder} itself
   */
  public setAlternativeAuthor(author: string): CommentBuilder {
    this.jsonMetadata.author = author;

    return this;
  }

  /**
   * Pushes images to the json metadata object
   *
   * @param {string} images image to be pushed to the json metadata object
   *
   * @returns {CommentBuilder} itself
   */
  public pushImages(...images: string[]): CommentBuilder {
    if(images.length === 0)
      return this;

    if(typeof this.jsonMetadata.image === "undefined")
      this.jsonMetadata.image = [];

    this.jsonMetadata.image.push(...images);

    return this;
  }

  /**
   * Pushes links to the json metadata object
   *
   * @param {string[]} links links to be pushed to the json metadata object
   *
   * @returns {CommentBuilder} itself
   */
  public pushLinks(...links: string[]): CommentBuilder {
    if(links.length === 0)
      return this;

    if(typeof this.jsonMetadata.links === "undefined")
      this.jsonMetadata.links = [];

    this.jsonMetadata.links.push(...links);

    return this;
  }

  /**
   * Sets description on the json metadata object
   *
   * @param {string} description description to be set on the json metadata object
   *
   * @returns {CommentBuilder} itself
   */
  public setDescription(description: string): CommentBuilder {
    this.jsonMetadata.description = description;

    return this;
  }

  /**
   * Sets format on the json metadata object
   *
   * @param {ECommentFormat} format format to be set on the json metadata object
   *
   * @returns {CommentBuilder} itself
   */
  public setFormat(format: ECommentFormat): CommentBuilder {
    this.jsonMetadata.format = format;

    return this;
  }

  /**
   * Adds beneficiary account to the comment operation object
   *
   * @param {TAccountName} account beneficiary account
   * @param {number} weight weight of the beneficiary account
   *
   * @returns {CommentBuilder} itself
   */
  public addBeneficiary(account: TAccountName, weight: number): CommentBuilder {
    this.ensureCommentOptionsCreated();

    let beneficiary: comment_payout_beneficiaries | undefined;

    for(const ext of this.commentOptions!.extensions)
      if("comment_payout_beneficiaries" in ext)
        beneficiary = ext.comment_payout_beneficiaries;


    if(typeof beneficiary === "undefined")
      this.commentOptions!.extensions.push({
        comment_payout_beneficiaries: { beneficiaries: [ { account, weight } ] }
      });
    else
      beneficiary.beneficiaries.push({ account, weight });

    return this;
  }

  /**
   * Adds beneficiary account(s) to the comment operation object
   *
   * @param {Array<beneficiary_route_type>} accounts beneficiary accounts
   *
   * @returns {CommentBuilder} itself
   */
  public addBeneficiaries(...accounts: Array<beneficiary_route_type>): CommentBuilder {
    for(const { account, weight } of accounts)
      this.addBeneficiary(account, weight);

    return this;
  }

  /**
   * Sets maximum accepted payout on the comment
   *
   * @param {NaiAsset} amount Maximum accepted payout (defaults to `1000000.000 HBD`)
   *
   * @returns {CommentBuilder} itself
   * @see {@link comment_options.max_accepted_payout}
   */
  public setMaxAcceptedPayout(amount: NaiAsset): CommentBuilder {
    this.ensureCommentOptionsCreated();

    this.commentOptions!.max_accepted_payout = amount;

    return this;
  }

  /**
   * Sets percent hbd
   *
   * @param {number} value percent hbd (defaults to `10000` - `HIVE_100_PERCENT`)
   *
   * @returns {CommentBuilder} itself
   * @see {@link comment_options.percent_hbd}
   */
  public setPercentHbd(value: number): CommentBuilder {
    this.ensureCommentOptionsCreated();

    this.commentOptions!.percent_hbd = value;

    return this;
  }

  /**
   * Sets allow curation rewards
   *
   * @param {boolean} value allow curation rewards value (defaults to `true`)
   *
   * @returns {CommentBuilder} itself
   * @see {@link comment_options.allow_curation_rewards}
   */
  public setAllowCurationRewards(value: boolean): CommentBuilder {
    this.ensureCommentOptionsCreated();

    this.commentOptions!.allow_curation_rewards = value;

    return this;
  }

  /**
   * Sets allow votes
   *
   * @param {boolean} value allow votes value (defaults to `true`)
   *
   * @returns {CommentBuilder} itself
   * @see {@link comment_options.allow_votes}
   */
  public setAllowVotes(value: boolean): CommentBuilder {
    this.ensureCommentOptionsCreated();

    this.commentOptions!.allow_votes = value;

    return this;
  }

  /**
   * @internal
   */
  public override build(): IBuiltHiveAppsOperation {
    // Apply cached json metadata before pushing the comment operation
    this.comment.json_metadata = JSON.stringify(this.jsonMetadata);

    this.push({ comment: this.comment });

    if(typeof this.commentOptions === "object")
      this.push({ comment_options: this.commentOptions });

    return this.builtOperations;
  }
}

/**
 * Same as the comment builder base, but requires parentAuthor and parentPermlink to be set
 */
export class ReplyBuilder extends CommentBuilder {
  public constructor(parentAuthor: TAccountName, parentPermlink: string, author: TAccountName, body: string, jsonMetadata?: object, permlink?: string, title: string = "") {
    super(parentAuthor, parentPermlink, author, body, jsonMetadata, permlink, title);

    if(parentAuthor.length === 0)
      throw new WaxError('No parent author specified in the reply builder');

    if(parentPermlink.length === 0)
      throw new WaxError('No parent permlink specified in the reply builder');
  }
}

/**
 * Same as the comment builder base, but requires user to set the category (parent permlink) on the comment
 */
export class ArticleBuilder extends CommentBuilder {
  public constructor(author: TAccountName, permlink: string, title: string, body: string, jsonMetadata?: object) {
    super('', '', author, body, jsonMetadata, permlink, title);
  }

  /**
   * Sets category of the article
   *
   * @param {string} category category (parent permlink) and also the first tag
   *
   * @returns {CommentBuilder} ready to build transaction builder
   */
  setCategory(category: string): CommentBuilder {
    this.comment.parent_permlink = category;

    return this;
  }
}

File: rust/wax/src/complex_operations/comment.rs
//! Builders for the `comment_operation` + optional `comment_options_operation`
//! pair, mirroring `ts/wasm/lib/detailed/complex_operations/comment.ts`.
//!
//! Two public builders: [`ReplyOperation`] (reply to an existing post/comment)
//! and [`BlogPostOperation`] (top-level post under a category). Both produce
//! a `comment_operation` and, if the caller set any option-shaped field, a
//! second `comment_options_operation`. The options op is suppressed when the
//! computed options equal the chain defaults — same behaviour as TS's
//! `deepEqual(defaultCommentOptions, this.commentOptions)` check.
//!
//! Differences from the TS port:
//! - `pushMetadataProperty` and the arbitrary user-supplied `jsonMetadata`
//!   object are not ported. Only the typed metadata fields (`format`,
//!   `tags`, `images`, `links`, `alternative_author`, `description`, `app`)
//!   are surfaced. `app` defaults to `wax/{CARGO_PKG_VERSION}` and can be
//!   overridden via the `app` field (mirrors TS's `jsonMetadata.app`).
//! - The default `comment_options` payload is hard-coded against the
//!   protocol constants rather than fetched through an FFI call to
//!   `cpp_get_default_comment_options_operation`. The values (1_000_000_000
//!   HBD payout, 10_000 bps percent_hbd, both `allow_*` true) are protocol-
//!   stable; this avoids adding bridge surface for a single helper.

use std::time::{SystemTime, UNIX_EPOCH};

use serde::Serialize;
use wax_core::proto;

use crate::WaxError;
use crate::constants::{
    DEFAULT_COMMENT_MAX_ACCEPTED_PAYOUT_SATOSHIS, DEFAULT_COMMENT_PERCENT_HBD,
};
use crate::foundation::WaxFoundation;
use crate::interfaces::OperationBuilder;
use crate::models::asset::{AssetName, NaiAsset, NaiAssetConvertible};
use crate::models::basic::AccountName;

const APP_TAG: &str =
    concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"));

/// Represents the wire-form `format` value written to `json_metadata.format`.
///
/// TS NOTE: mirrors `ECommentFormat`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CommentFormat {
    Html,
    Markdown,
    Mixed,
}

impl CommentFormat {
    fn as_str(self) -> &'static str {
        match self {
            CommentFormat::Html => "html",
            CommentFormat::Markdown => "markdown",
            CommentFormat::Mixed => "markdown+html",
        }
    }
}

/// Represents one beneficiary entry for `comment_options.extensions`, mirroring
/// `proto::BeneficiaryRouteType` without forcing callers to import from
/// `wax::proto`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BeneficiaryRoute {
    pub account: AccountName,
    pub weight: u32,
}

/// Represents a reply / comment under an existing post or comment.
/// `parent_author` and `parent_permlink` are required and non-empty; an empty
/// value for either is rejected at `finalize` time.
#[derive(Debug, Clone, Default)]
pub struct ReplyOperation {
    pub parent_author: AccountName,
    pub parent_permlink: String,
    pub author: AccountName,
    pub body: String,
    /// Defaults to `"re-{parent_author}-{millis}"` if `None`.
    pub permlink: Option<String>,
    /// Defaults to `""` if `None`.
    pub title: Option<String>,

    pub format: Option<CommentFormat>,
    pub tags: Vec<String>,
    pub images: Vec<String>,
    pub links: Vec<String>,
    pub alternative_author: Option<AccountName>,
    pub description: Option<String>,
    /// Overrides the `app` tag in `json_metadata`. Defaults to
    /// `wax/{CARGO_PKG_VERSION}` when `None`.
    pub app: Option<String>,

    pub beneficiaries: Vec<BeneficiaryRoute>,
    pub allow_curation_rewards: Option<bool>,
    pub allow_votes: Option<bool>,
    pub percent_hbd: Option<u32>,
    pub max_accepted_payout: Option<NaiAssetConvertible>,
}

/// Represents a top-level blog post under a category (e.g. `"travel"` or a
/// community id like `"hive-174695"`). `parent_author` is always `""`, and
/// `parent_permlink = category`.
#[derive(Debug, Clone, Default)]
pub struct BlogPostOperation {
    pub author: AccountName,
    pub category: String,
    pub title: String,
    pub body: String,
    /// Defaults to `"{author}-{millis}"` if `None`.
    pub permlink: Option<String>,

    pub format: Option<CommentFormat>,
    pub tags: Vec<String>,
    pub images: Vec<String>,
    pub links: Vec<String>,
    pub alternative_author: Option<AccountName>,
    pub description: Option<String>,
    /// Overrides the `app` tag in `json_metadata`. Defaults to
    /// `wax/{CARGO_PKG_VERSION}` when `None`.
    pub app: Option<String>,

    pub beneficiaries: Vec<BeneficiaryRoute>,
    pub allow_curation_rewards: Option<bool>,
    pub allow_votes: Option<bool>,
    pub percent_hbd: Option<u32>,
    pub max_accepted_payout: Option<NaiAssetConvertible>,
}

/// Internal merged representation used by both builders' `finalize` paths.
/// Carries the fully-resolved `parent_*` / `permlink` / `title` values so
/// the shared finalize logic doesn't have to know which builder produced it.
struct CommentInputs {
    parent_author: String,
    parent_permlink: String,
    author: String,
    permlink: String,
    title: String,
    body: String,
    format: Option<CommentFormat>,
    tags: Vec<String>,
    images: Vec<String>,
    links: Vec<String>,
    alternative_author: Option<AccountName>,
    description: Option<String>,
    app: Option<String>,
    beneficiaries: Vec<BeneficiaryRoute>,
    allow_curation_rewards: Option<bool>,
    allow_votes: Option<bool>,
    percent_hbd: Option<u32>,
    max_accepted_payout: Option<NaiAssetConvertible>,
}

impl OperationBuilder for ReplyOperation {
    fn finalize(
        self: Box<Self>,
        foundation: &dyn WaxFoundation,
    ) -> Result<Vec<proto::Operation>, WaxError> {
        let this = *self;
        if this.parent_author.is_empty() {
            return Err(WaxError::new(
                "No parent author specified in the reply operation",
            ));
        }
        if this.parent_permlink.is_empty() {
            return Err(WaxError::new(
                "No parent permlink specified in the reply operation",
            ));
        }

        let permlink = this.permlink.unwrap_or_else(|| {
            format!("re-{}-{}", this.parent_author, now_millis())
        });

        let inputs = CommentInputs {
            parent_author: this.parent_author,
            parent_permlink: this.parent_permlink,
            author: this.author,
            permlink,
            title: this.title.unwrap_or_default(),
            body: this.body,
            format: this.format,
            tags: this.tags,
            images: this.images,
            links: this.links,
            alternative_author: this.alternative_author,
            description: this.description,
            app: this.app,
            beneficiaries: this.beneficiaries,
            allow_curation_rewards: this.allow_curation_rewards,
            allow_votes: this.allow_votes,
            percent_hbd: this.percent_hbd,
            max_accepted_payout: this.max_accepted_payout,
        };
        inputs.into_operations(foundation)
    }
}

impl OperationBuilder for BlogPostOperation {
    fn finalize(
        self: Box<Self>,
        foundation: &dyn WaxFoundation,
    ) -> Result<Vec<proto::Operation>, WaxError> {
        let this = *self;
        let permlink = this
            .permlink
            .unwrap_or_else(|| format!("{}-{}", this.author, now_millis()));

        let inputs = CommentInputs {
            parent_author: String::new(),
            parent_permlink: this.category,
            author: this.author,
            permlink,
            title: this.title,
            body: this.body,
            format: this.format,
            tags: this.tags,
            images: this.images,
            links: this.links,
            alternative_author: this.alternative_author,
            description: this.description,
            app: this.app,
            beneficiaries: this.beneficiaries,
            allow_curation_rewards: this.allow_curation_rewards,
            allow_votes: this.allow_votes,
            percent_hbd: this.percent_hbd,
            max_accepted_payout: this.max_accepted_payout,
        };
        inputs.into_operations(foundation)
    }
}

impl CommentInputs {
    /// `true` iff the caller touched any option-shaped field. Triggers
    /// lazy creation of the `comment_options_operation` per TS semantics.
    fn has_options_input(&self) -> bool {
        self.allow_curation_rewards.is_some()
            || self.allow_votes.is_some()
            || self.percent_hbd.is_some()
            || self.max_accepted_payout.is_some()
            || !self.beneficiaries.is_empty()
    }

    fn into_operations(
        self,
        foundation: &dyn WaxFoundation,
    ) -> Result<Vec<proto::Operation>, WaxError> {
        // Snapshot the options-trigger before consuming `self` field-by-field
        // below — `has_options_input` needs `&self` and would conflict with
        // the partial moves into `proto::Comment`.
        let has_options = self.has_options_input();

        let json_metadata = build_json_metadata(
            self.format.unwrap_or(CommentFormat::Mixed),
            &self.tags,
            &self.images,
            &self.links,
            self.alternative_author.as_deref(),
            self.description.as_deref(),
            self.app.as_deref(),
        )?;

        let comment_op = proto::Comment {
            parent_author: self.parent_author,
            parent_permlink: self.parent_permlink,
            author: self.author.clone(),
            permlink: self.permlink.clone(),
            title: self.title,
            body: self.body,
            json_metadata,
        };

        let mut ops = Vec::with_capacity(2);
        ops.push(proto::Operation {
            value: Some(proto::operation::Value::CommentOperation(comment_op)),
        });

        if has_options {
            let default = foundation
                .default_comment_options(&self.author, &self.permlink)?;
            let computed = build_comment_options(
                foundation,
                &self.author,
                &self.permlink,
                self.allow_curation_rewards,
                self.allow_votes,
                self.percent_hbd,
                self.max_accepted_payout,
                &self.beneficiaries,
            )?;

            if computed != default {
                ops.push(proto::Operation {
                    value: Some(
                        proto::operation::Value::CommentOptionsOperation(
                            computed,
                        ),
                    ),
                });
            }
        }

        Ok(ops)
    }
}

fn now_millis() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("system clock before unix epoch")
        .as_millis() as u64
}

/// Internal struct backing the JSON metadata serialization. Field
/// declaration order matters: `serde_json` preserves it, so the output
/// matches the TS `JSON.stringify` shape byte-for-byte
/// (`format`, `app`, `tags`, `image`, `links`, `author`, `description`).
#[derive(Serialize)]
struct JsonMetadata<'a> {
    format: &'a str,
    app: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    image: Option<&'a [String]>,
    #[serde(skip_serializing_if = "Option::is_none")]
    links: Option<&'a [String]>,
    #[serde(skip_serializing_if = "Option::is_none")]
    author: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<&'a str>,
}

fn build_json_metadata(
    format: CommentFormat,
    tags: &[String],
    images: &[String],
    links: &[String],
    alternative_author: Option<&str>,
    description: Option<&str>,
    app: Option<&str>,
) -> Result<String, WaxError> {
    let metadata = JsonMetadata {
        format: format.as_str(),
        app: app.unwrap_or(APP_TAG),
        tags: if tags.is_empty() {
            None
        } else {
            Some(deduplicate_preserving_order(tags))
        },
        image: if images.is_empty() {
            None
        } else {
            Some(images)
        },
        links: if links.is_empty() { None } else { Some(links) },
        author: alternative_author,
        description,
    };

    serde_json::to_string(&metadata).map_err(|e| {
        WaxError::new(format!("failed to serialize comment json_metadata: {e}"))
    })
}

/// First-occurrence-wins dedupe preserving order. Mirrors TS
/// `[...new Set([...existing, ...data.tags])]`.
fn deduplicate_preserving_order(items: &[String]) -> Vec<String> {
    let mut seen = std::collections::HashSet::with_capacity(items.len());
    let mut out = Vec::with_capacity(items.len());
    for item in items {
        if seen.insert(item.clone()) {
            out.push(item.clone());
        }
    }
    out
}

#[allow(clippy::too_many_arguments)]
fn build_comment_options(
    foundation: &dyn WaxFoundation,
    author: &str,
    permlink: &str,
    allow_curation_rewards: Option<bool>,
    allow_votes: Option<bool>,
    percent_hbd: Option<u32>,
    max_accepted_payout: Option<NaiAssetConvertible>,
    beneficiaries: &[BeneficiaryRoute],
) -> Result<proto::CommentOptions, WaxError> {
    let max_accepted_payout: NaiAsset = match max_accepted_payout {
        Some(asset) => foundation
            .create_asset_with_required_symbol(AssetName::Hbd, asset)?,
        None => foundation
            .hbd_satoshis(DEFAULT_COMMENT_MAX_ACCEPTED_PAYOUT_SATOSHIS)?,
    };

    let extensions = if beneficiaries.is_empty() {
        Vec::new()
    } else {
        vec![proto::CommentOptionsExtension {
            value: Some(
                proto::comment_options_extension::Value::CommentPayoutBeneficiaries(
                    proto::CommentPayoutBeneficiaries {
                        beneficiaries: beneficiaries
                            .iter()
                            .map(|b| proto::BeneficiaryRouteType {
                                account: b.account.clone(),
                                weight: b.weight,
                            })
                            .collect(),
                    },
                ),
            ),
        }]
    };

    Ok(proto::CommentOptions {
        author: author.to_string(),
        permlink: permlink.to_string(),
        max_accepted_payout,
        percent_hbd: percent_hbd.unwrap_or(DEFAULT_COMMENT_PERCENT_HBD),
        allow_votes: allow_votes.unwrap_or(true),
        allow_curation_rewards: allow_curation_rewards.unwrap_or(true),
        extensions,
    })
}
